standards/development-principles.md

```diff
@@ -0,0 +1,93 @@
+Don’t Repeat Yourself - The ‘[Rule of Three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming))’ is a good approach to managing duplication. Less code is usually better - but not at the expense of clarity
+
+### 3. Optimize for change
+Don’t try to solve every conceivable problem up-front, instead focus on making your code easy to change when needed
+Don’t prematurely optimize - choose clarity over performance, unless there is a serious performance issue that needs to be addressed.
+
+### 4. Something simple which exists is better than a perfect solution which doesn’t exist yet
+- We learn best from putting real things in front of real users, not trying to achieve perfection in isolation. There is no substitute for a real user clicking through a user journey in a real application. Get it done, get it in front of people, and learn early.
+
+- Don’t over-engineer. Follow the principle of Least Surprise. Choose process
+concurrency over threading & let the O/S handle it, unless there is good reason.
+When you do use threading, use language abstractions to help. Don’t roll your
+own crypto. Handle exceptions at the app level, no lower. Etc etc
+
+### 5. Everything fails, all of the time.
+Accept this and [code defensively](https://en.wikipedia.org/wiki/Defensive_programming) when calling other services.
+Every HTTP call could error or hang - handle failures appropriately and fail fast. Don’t let long-running external calls impact your user experience.
+
+### 6. Other developers are users too - treat them with respect
+If you have to explain how your code works, then your code is not clear enough.
+Comments are for explaining <strong>why</strong> something is needed, not how it works.
+Commit messages should follow [GDS guidance](https://github.com/alphagov/styleguides/blob/master/git.md)
+No-one cares how clever you are - it’s far more important to work well with your team
+APIs are interfaces too - Like any other interface, APIs need designing and iterating for usability
+Don’t pollute the global namespace
+
+### 7. Think smaller
+Stick to the Single Responsibility Principle - keep view templates, controllers and model classes as simple as possible. Keep methods short. (Concepts and patterns which may help with this: Null Object pattern, Facade pattern, Form Objects, Sandi Metz’s “[Rules for Developers](https://robots.thoughtbot.com/sandi-metz-rules-for-developers)”)
+A solution composed of many small simple things is usually better than one big complex thing.
+
+### 8. Names have power. Use them wisely.
+Don’t be cute or jokey when naming things
+Names convey meaning - well-named functions & variables can remove the need for a comment
+Avoid meaningless names like ‘obj’ / ‘result’ / ‘foo’.
+Use single-letter variables only where the letter represents a well-known mathematical property (e.g. e = mc^2), or where their meaning is otherwise clear
```

Comment:
This section is already included and expanded on in the [page about naming things](https://github.com/ministryofjustice/technical-guidance/blob/442de9c5305202f3e912a9dc02dad4301fa37846/standards/naming-things.md).

Reply:
Naming things seems to touch on naming services where this text is referring to functions and variables etc. but lets discuss this further once this is out there. I have added a link to about naming things so that we can decided which one is better placed in this context.